Split adaptation code to backend and add libgbinder based backend. MER#1942

mlehtima opened the pull request:
Not all extension functions and callbacks have yet been fully tested. Requires https://github.com/mer-hybris/libgbinder/pull/4

monich commented on binder/binder.pro:
```diff
@@ -0,0 +1,19 @@
+target.path = /usr/libexec
+INSTALLS += target
+
+CONFIG += install_ok  # Do not cargo-cult this!
```
Do not cargo-cult this :)  I don't think this `install_ok` thing is necessary.

mlehtima replied:
I used the Qt example as the base for the project file, I'll remove unneeded lines.

monich commented on hybrisprovider.cpp:
```diff
@@ -593,74 +256,41 @@ HybrisProvider::HybrisProvider(QObject *parent)
+    if (plugin) {
+        m_backend = qobject_cast<HybrisBackend *>(plugin);
+    } else {
+        qWarning(pluginLoader.errorString().toStdString().c_str());
```
`qPrintable`, maybe? Or maybe not, it's fine either way.

mlehtima replied:
I'll have a look if this can be made simpler.

monich commented on hybrisprovider.cpp:
```diff
@@ -593,74 +256,41 @@ HybrisProvider::HybrisProvider(QObject *parent)
+    QDir pluginsDir(qApp->applicationDirPath());
+    bool binderBackendFound = false;
+    foreach (QString fileName, pluginsDir.entryList(QDir::Files)) {
+        if (fileName.contains("libgeoclue-hybris-binder.so")) {
```
Could plugin selection and loading be slightly more generic? Like, check all libraries with certain prefix and pick the first one that implements the required interface, or something like that. Do we really need to support scenario with multiple plugins installed?

mlehtima replied:
We always have the hal plugin present so we want to prefer the binder plugin. If we try to use the first one found then we would have to try to initialize the plugin and if initializing fails then continue to the next one found. This is possible if that is wanted.

monich replied:
Ok, so multiple plugins is a realistic and apparently meaningful scenario.

If it's hal plugin that for whatever reason has to be always present (sounds strange to me but suppose it does make sense) then I suggest that the logic should be slightly different - look for any plugin other than hal, if it's found then use that, otherwise use the default hal thing.

In other words, hardcode as few file names as possible (only `libgeoclue-hybris-hal.so`).

mlehtima replied:
Presence of HAL plugin is for backwards compatibility to make sure building only the default package would be enough in old adaptations. Other option would be to add some suitable dependencies in each package so by default the HAL package would be used but if some other package is forcefully installed then the HAL plugin package would not be installed at all and we would have one plugin installed at all times.

monich replied:
Yes, that's kind of like what I had in mind. Provider's package may require something (e.g. `geoclue-providers-backend`) and backend packages provide it. Then it's a matter of which backend package gets pulled in by the build. That would naturally make backends exclusive, and we won't have to bother with multiple backends being installed simultaneously. I'm not quite sure though, how to make it most convenient for porters (i.e. how to automatically provide a reasonable default without any additional work being done), that's something you probably know better than I.

monich commented:
mer-hybris/libgbinder#4 has been merged, by the way.

chriadam commented on binder/binder.cpp:
```diff
@@ -0,0 +1,1150 @@
+            }
+            }
+            break;
+        case GNSS_ACQUIRE_WAKELOCK_CB:
```
Should these call nemo-keepalive etc, or are these merely informative that linux wakelocks have been aquired/released?

mlehtima replied:
These were never used in the HAL implementation so I ignored those as was done before The description in the documentation is "Callback utility for acquiring the GNSS wakelock. This can be used to prevent the CPU from suspending while handling GNSS events." so nemo-keepalive could be used. Not sure why it wasn't used in the old HAL implementation.

chriadam commented on binder/binder.cpp:
```diff
@@ -0,0 +1,1150 @@
+            qCDebug(lcGeoclueHybris) << "GNSS request UTC time";
+            QMetaObject::invokeMethod(staticProvider, "injectUtcTime", Qt::QueuedConnection);
+            break;
+        case GNSS_SET_SYSTEM_INFO_CB:
```
ooi what is this intended for?

mlehtima replied:
Currently system info callback only provides "year in which the last update was made to the underlying hardware/firmware used to capture GNSS signals" which is not very relevant for us. We could log it but otherwise we don't have any use for it. The year is used in Android CTS tests to decide which features are required.

chriadam commented on binder/binder.cpp:
```diff
@@ -0,0 +1,1150 @@
+
+HybrisBinder::HybrisBinder(QObject *parent)
+:   QObject(parent), m_death_id(0), m_fqname(0), m_sm(0),
+    m_clientGnss(0),
```
nitpick: could use Q_NULLPTR instead of literal 0 for pointer initialization

mlehtima replied:
The original code used 0 for null pointer so I used the same to make the code consistent. Maybe we should replace similar values also in the old code with Q_NULLPTR?

chriadam commented:
Fantastic work, thanks very much.

chriadam commented on binder/binder.cpp:
```diff
@@ -0,0 +1,1150 @@
+                SatelliteInfo satInfo;
+                GnssSvInfo svInfo = svStatus->gnssSvList.data[i];
+                satInfo.setPrn(svInfo.svid);
+                satInfo.setSnr(svInfo.cN0Dbhz);
```
I guess the signal to noise ratio is usually a value from 0.0 to 1.0, but the carrier/noise value may not be?  Should this be normalised somehow to a valid snr value?  (I checked https://doc-snapshots.qt.io/qt5-5.11/qgeosatelliteinfo.html#signalStrength and it seems that this isn't documented to be SNR specifically, and in fact has no documented semantics at all, which is unfortunate.  But value-range consistency would be appreciated nonetheless.)

mlehtima replied:
I will have to check how the satellite info in handled in other components and decide how to handle that.

mlehtima replied:
I checked the values we get from the snr field in older adaptations and I saw on FP2 at least values between 0-40 so I think we can still use the raw value we get from the cN0Dbhz field of GnssSvInfo.

mlehtima commented:
Code has been updated with null pointers now using Q_NULLPTR, backends are now built-in and both spec files build a complete binary with only one backend so only one package is used at all times, fixed and cleaned up various build and dependency related issues, fixed satellite IDs (prn) reported by the new satellite info callbacks to match the old API so QtLocation users can determine the satellite system from those values (previously values had overlap which caused issues).

mlehtima commented:
Fixed a memory leak and some style inconsistencies.
